staging/vhost-device-gpu/src/virtio_gpu.rs

```diff
@@ -0,0 +1,1055 @@
+    pub(crate) rutabaga: Rutabaga,
+    pub(crate) resources: BTreeMap<u32, VirtioGpuResource>,
+    pub(crate) fence_state: Arc<Mutex<FenceState>>,
+    pub(crate) scanouts: [Option<VirtioGpuScanout>; VIRTIO_GPU_MAX_SCANOUTS],
+}
+
+const READ_RESOURCE_BYTES_PER_PIXEL: usize = 4;
+
+impl RutabagaVirtioGpu {
+    // TODO: this depends on Rutabaga builder, so this will need to be handled at runtime eventually
+    pub const MAX_NUMBER_OF_CAPSETS: u32 = 3;
+
+    fn create_fence_handler(
+        queue_ctl: VringRwLock,
+        fence_state: Arc<Mutex<FenceState>>,
+    ) -> RutabagaFenceHandler {
+        RutabagaFenceHandler::new(move |completed_fence: RutabagaFence| {
+            debug!(
+                "XXX - fence called: id={}, ring_idx={}",
+                completed_fence.fence_id, completed_fence.ring_idx
+            );
+
+            //let mut queue = queue_ctl.lock().unwrap();
+            let mut fence_state = fence_state.lock().unwrap();
+            let mut i = 0;
+
+            let ring = match completed_fence.flags & VIRTIO_GPU_FLAG_INFO_RING_IDX {
+                0 => VirtioGpuRing::Global,
+                _ => VirtioGpuRing::ContextSpecific {
+                    ctx_id: completed_fence.ctx_id,
+                    ring_idx: completed_fence.ring_idx,
+                },
+            };
+
+            while i < fence_state.descs.len() {
+                debug!("XXX - fence_id: {}", fence_state.descs[i].fence_id);
+                if fence_state.descs[i].ring == ring
+                    && fence_state.descs[i].fence_id <= completed_fence.fence_id
+                {
+                    let completed_desc = fence_state.descs.remove(i);
+                    debug!(
+                        "XXX - found fence: desc_index={}",
+                        completed_desc.desc_index
+                    );
+
+                    queue_ctl
+                        .add_used(completed_desc.desc_index, completed_desc.len)
+                        .unwrap();
+
+                    queue_ctl
+                        .signal_used_queue()
+                        .map_err(Error::NotificationFailed)
+                        .unwrap();
+                    debug!("Notification sent");
+                } else {
+                    i += 1;
+                }
+            }
+            // Update the last completed fence for this context
+            fence_state
+                .completed_fences
+                .insert(ring, completed_fence.fence_id);
+        })
+    }
+
+    pub fn new(queue_ctl: &VringRwLock) -> Self {
+        let xdg_runtime_dir = match env::var("XDG_RUNTIME_DIR") {
+            Ok(dir) => dir,
+            Err(_) => "/run/user/1000".to_string(),
+        };
+        let wayland_display = match env::var("WAYLAND_DISPLAY") {
+            Ok(display) => display,
+            Err(_) => "wayland-0".to_string(),
+        };
+        let path = PathBuf::from(format!("{}/{}", xdg_runtime_dir, wayland_display));
```

Comment:
This does not seem right (especially the hardcoded `/run/user/1000` part). Looking at [some wayland docs](https://wayland.freedesktop.org/docs/html/apb.html#Client-classwl__display_1af048371dfef7577bd39a3c04b78d0374), I found this:

> 
>  Connect to the Wayland display named name. If name is NULL, its value will be replaced with the WAYLAND_DISPLAY environment variable if it is set, otherwise display "wayland-0" will be used.
> 
> If WAYLAND_SOCKET is set, it's interpreted as a file descriptor number referring to an already opened socket. In this case, the socket is used as-is and name is ignored.
> 
> If name is a relative path, then the socket is opened relative to the XDG_RUNTIME_DIR directory.
> 
> If name is an absolute path, then that path is used as-is for the location of the socket at which the Wayland server is listening; no qualification inside XDG_RUNTIME_DIR is attempted.
> 
> If name is NULL and the WAYLAND_DISPLAY environment variable is set to an absolute pathname, then that pathname is used as-is for the socket in the same manner as if name held an absolute path. Support for absolute paths in name and WAYLAND_DISPLAY is present since Wayland version 1.15.

Reply:
There's no "typically" here [if the wayland docs say otherwise](https://wayland.freedesktop.org/docs/html/apb.html#Client-classwl__display), right? Also no, uid is not always `1000` and not always under `XDG_RUNTIME_DIR`.

Reply:
ok thanks for the pointer, you don't need to be rude though. I will change it removing the Err match case. variant so as to not assume a WAYLAND_DISPLAY in case of Err.

Reply:
Oh no it wasn't meant to be rude at all! I'm so sorry :( honestly. I was pointing out that this is the "correct" way so you can benefit from it, I did not mean to write it in such a way.

Reply:
ok, I understand. no worries then.

Reply:
In the future if there's any problem please reach out in private first. This is a public professional setting.

Reply:
About this, I removed `.set_rutabaga_channels(rutabaga_channels_opt)` because it's not necessary for our use case. this is related to running sommerlier, like [crosvm](https://crosvm.dev/book/devices/wayland.html?highlight=gpu#running-guest-wayland-apps) does, and this is for crossdomain component which we do not support in our vhost-device-gpu.